### PR TITLE
Implement cross-compilation support Linux → Darwin

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Build/Testing/Packaging Improvement

### CODE REVIEW

File: cmake/target.cmake [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+if (CMAKE_CROSSCOMPILING)
+    if (CMAKE_SYSTEM_NAME MATCHES "Darwin")
+        set (CMAKE_SYSTEM_PROCESSOR x86_64)
+        set (CMAKE_C_COMPILER_TARGET x86_64-apple-darwin)
+        set (CMAKE_CXX_COMPILER_TARGET x86_64-apple-darwin)
+
+        set (HAS_PRE_1970_EXITCODE "0" CACHE STRING "Result from TRY_RUN" FORCE)
+        set (HAS_PRE_1970_EXITCODE__TRYRUN_OUTPUT "" CACHE STRING "Output from TRY_RUN" FORCE)
+
+        set( HAS_POST_2038_EXITCODE "0" CACHE STRING "Result from TRY_RUN" FORCE)
```
Comment (alexey-milovidov) [inline cmake/target.cmake]: Inconsistent parens.

File: cmake/target.cmake [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    set (CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -fuse-ld=${LINKER_NAME}")
+endif ()
+
+if (CMAKE_CROSSCOMPILING)
+    if (CMAKE_SYSTEM_NAME MATCHES "Darwin")
+        set (CMAKE_SYSTEM_PROCESSOR x86_64)
+        set (CMAKE_C_COMPILER_TARGET x86_64-apple-darwin)
+        set (CMAKE_CXX_COMPILER_TARGET x86_64-apple-darwin)
+
+        set (HAS_PRE_1970_EXITCODE "0" CACHE STRING "Result from TRY_RUN" FORCE)
```
Comment (alexey-milovidov) [inline cmake/target.cmake]: What does it mean?

Comment (abyss7) [reply cmake/target.cmake]: One of our contribs requires this variables to be set. The values are results of running some binary, which has mach-o format in cross-build. CMake explicitly asks to run this binary on MacOS and set the values to cached variables.

File: cmake/target.cmake (hunk 1 above)
Comment (abyss7) [reply cmake/target.cmake]: Fixed

File: cmake/find/snappy.cmake [hunk 3]
```diff
@@ -0,0 +1,13 @@
+option(USE_SNAPPY "Enable support of snappy library" ON)
```
Comment (proller) [inline cmake/find/snappy.cmake]: Please keep MISSING_INTERNAL_SNAPPY_LIBRARY check

Comment (abyss7) [reply cmake/find/snappy.cmake]: Let's put this and all similar discussions about auto-detection and related features into #6981 - for historical references in the future.

File: dbms/src/Functions/registerFunctionsIntrospection.cpp [hunk 4]
```diff
@@ -1,9 +1,11 @@
+#include <Core/Defines.h>
+
 namespace DB
 {
 
 class FunctionFactory;
 
-#ifdef __ELF__
+#if defined(OS_LINUX)
```
Comment (alexey-milovidov) [inline dbms/src/Functions/registerFunctionsIntrospection.cpp]: But these functions can work on ELF based systems such as FreeBSD.

**Comment by alexey-milovidov** [comment]
Up next:
- a command to download SDK in build instruction;
- add to CI and provide tarball to download.

File: dbms/src/Functions/registerFunctionsIntrospection.cpp (hunk 4 above)
Comment (abyss7) [reply dbms/src/Functions/registerFunctionsIntrospection.cpp]: Reasonable.

**Pull request merged**